src/sumtype.ts

```diff
@@ -19,35 +19,35 @@ export type CasePattern<T extends Variants, R> =
   | Partial<ExhaustiveCasePattern<T, R>> & { _: () => R };
 
 abstract class SumType<M extends Variants> implements Setoid, Show {
-  private kind: keyof M;
-  private data: unknown[];
+  #kind: keyof M;
+  #data: unknown[];
 
   constructor(...args: KindAndData<M>) {
     let [kind, ...data] = args;
-    this.kind = kind;
-    this.data = data;
+    this.#kind = kind;
+    this.#data = data;
   }
 
   public caseOf<T>(pattern: CasePattern<M, T>): T {
-    if (this.kind in pattern) {
-      return (pattern[this.kind] as any)(...this.data);
+    if (this.#kind in pattern) {
+      return (pattern[this.#kind] as any)(...this.#data);
     } else if (pattern._) {
       return pattern._();
     } else {
-      throw new Error(`caseOf pattern is missing a function for ${this.kind}`);
+      throw new Error(`caseOf pattern is missing a function for ${this.#kind}`);
     }
   }
 
   public equals(that: SumType<M>): boolean {
-    return this.kind === that.kind && arrayEquals(this.data, that.data);
+    return this.toString() === that.toString();
```

Comment:
This didn't feel great, but not sure if there's a better alternative.

Reply:
As I understand it, ES private names are effectively lexically scoped, so you should be able to keep the previous implementation if you sprinkle some `#`s in there

Reply:
I wouldn't be able to get it on `that` :(

Reply:
Are you sure? My understanding from the proposal text was that you can access private fields on any instance of the class as long as that access occurred within the body of the class declaration.

Reply:
http://www.typescriptlang.org/play#code/MYGwhgzhAEDCCmIQB4AqA+aBvAUNf0AxAG5ggCu8AXNKgNw54EAO5ARiAJbDTAD2AOwgAXAE7lgwvqIAUpCtVoBKbEwIFhAC04QAdCTKVoAXmjzKDdQF9G61h27R4AR3JkIMvlviiaCJGjoSjRsfHwg8GACqurqovDC5KLRWjr65vAmxqZemj7phvCWBDZWQA

Reply:
Wow thats super strange to me.

Reply:
Updated :)